src/backend/distributed/master/master_stage_protocol.c

```diff
@@ -562,7 +562,37 @@ CreateShardsOnWorkersViaExecutor(Oid distributedRelationId, List *shardPlacement
 		taskList = lappend(taskList, task);
 	}
 
-	ExecuteTaskList(CMD_UTILITY, taskList, poolSize);
+	if (useExclusiveConnection)
+	{
+		/*
+		 * When the table has local data, we force max parallelization so data
+		 * copy is done efficiently.
+		 */
+		ForceMaxQueryParallelization = true;
+
+		/*
+		 * TODO: we shouldn't need to set poolSize when setting ForceMaxQueryParallelization.
+		 *
+		 * Adaptive executor uses poolSize to decide if it should record parallel
+		 * access to relations or not, and it ignores ForceMaxQueryParallelization
+		 * because of some complications in TRUNCATE.
+		 */
+		poolSize = MaxAdaptiveExecutorPoolSize;
+	}
+
+	PG_TRY();
+	{
+		ExecuteTaskList(CMD_UTILITY, taskList, poolSize);
+	}
+	PG_CATCH();
+	{
+		ForceMaxQueryParallelization = prevForceMaxQueryParallelization;
```

Comment:
I think we can avoid the PG_TRY by using set_config_option (SET LOCAL), see SetLocalMultiShardModifyModeToSequential.

Reply:
Maybe double check that `useExclusiveConnection ` is also set true if `StoredProcedureLevel>1` or `FunctionCallLevel>1`. Can we safely use set local in those cases?

Reply:
> Maybe double check that `useExclusiveConnection ` is also set true if `StoredProcedureLevel>1` or `FunctionCallLevel>1`. Can we safely use set local in those cases?

I tried `SET LOCAL` in stored procedures and functions and they work as expected.